atlasdb-autobatch/src/main/java/com/palantir/atlasdb/autobatch/CoalescingBatchingEventHandler.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.autobatch;
+
+import java.util.Map;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.SetMultimap;
+import com.google.common.util.concurrent.SettableFuture;
+import com.lmax.disruptor.EventHandler;
+
+final class CoalescingBatchingEventHandler<T, R> implements EventHandler<BatchElement<T, R>> {
+
+    private final CoalescingRequestFunction<T, R> function;
+    private final SetMultimap<T, SettableFuture<R>> pending;
+    private final R defaultValue;
```

Comment:
Would be nice to have some javadoc overall on what this `defaultValue` is intended for - I guess this is done defensively, to avoid threads hanging infinitely should the batch function accidentally not return all keys provided? It seems like it would be enough to demand as a precondition that the map produced must have all keys from the input is sufficient.

Reply:
yeah, I was back and forth on this, I can change this

Reply:
will keep for now, might change later depending on how problematic this is

Reply:
discussed offline - this is relatively cheap to change, and we don't necessarily want to force that precondition demand